AIP-7393 Add Flow labels to Workflow for ArgoUI

talebzeghmi opened the pull request:
please see run with Workflow labels here:

* run: `python metaflow/tutorials/00-helloworld/helloworld.py aip run --tag t42 -e e42`
* Argo UIhttps://argo-server.int.stage-k8s.zg-aip.net/argo-ui/workflows/aip-playground-dev/helloflow-nbbdj?tab=summary
* Related Metaflow UI https://metaflow.int.stage-k8s.zg-aip.net/HelloFlow/argo-helloflow-nbbdj


```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
...
  labels:
    metaflow.org/experiment: e42
    metaflow.org/flow_name: HelloFlow
    metaflow.org/tag_e42: "true"
    metaflow.org/tag_t42: "true"
    pipelines.kubeflow.org/kfp_sdk_version: 1.0.81-1.8.10
    workflows.argoproj.io/phase: Running
    zodiac.zillowgroup.net/owner: talebz
```

and now searchable in the Argo UI labels filter

<img width="505" alt="image" src="https://github.com/zillow/metaflow/assets/4167032/f2f0db10-2ef2-4d0c-bcb2-80399f169522">


## compiled
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: helloflow-
  labels:
    pipelines.kubeflow.org/kfp_sdk_version: 1.0.81-1.8.10
    #  ↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓  PLEASE SEE LABELS created by this PR ↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓ 
    metaflow.org/flow_name: HelloFlow
    metaflow.org/experiment: e42
    metaflow.org/tag_t42: "true"
    zodiac.zillowgroup.net/owner: talebz
spec:
  entrypoint: helloflow
  templates:
  - name: end
    # existing labels and annotations added by WFSDK, and kfp
    metadata:
      annotations: 
        metaflow.org/step: end
        metaflow.org/run_id: 'argo-{{workflow.name}}'
        pipelines.kubeflow.org/task_display_name: end, 
        pipelines.kubeflow.org/max_cache_staleness: P0D
      labels:
        pipelines.kubeflow.org/kfp_sdk_version: 1.0.81-1.8.10
        pipelines.kubeflow.org/pipeline-sdk-type: kfp
        aip.zillowgroup.net/aip-pod-default: "true"
        sidecar.istio.io/inject: "false"
        metaflow.org/flow_name: HelloFlow
        zodiac.zillowgroup.net/owner: talebz
        tags.ledger.zgtools.net/ai-flow-name: HelloFlow
        tags.ledger.zgtools.net/ai-step-name: end
        pipelines.kubeflow.org/enable_caching: "true"
 ```

## k8s cluster podspec after AIP webhook and KFP mutations
```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    karpenter.sh/do-not-evict: "true"
    kubectl.kubernetes.io/default-container: main
    kubernetes.io/limit-ranger: 'LimitRanger plugin set: ephemeral-storage request
      for container wait; ephemeral-storage limit for container wait; cpu, ephemeral-storage,
      memory request for container main; cpu, ephemeral-storage, memory limit for
      container main; ephemeral-storage request for init container init; ephemeral-storage
      limit for init container init'
    kubernetes.io/psp: eks.privileged
    logging.zgtools.net/index: ai_platform_nonprod
    logging.zgtools.net/istio-init.ignore: "true"
    logging.zgtools.net/istio-proxy.ignore: "true"
    logging.zgtools.net/queue-proxy.ignore: "true"
    logging.zgtools.net/wait.ignore: "true"
    metaflow.org/run_id: argo-helloflow-8nvp9
    metaflow.org/step: start
    pipelines.kubeflow.org/max_cache_staleness: P0D
    pipelines.kubeflow.org/task_display_name: start
    sidecar.istio.io/inject: "false"
    workflows.argoproj.io/node-id: helloflow-8nvp9-1794368775
    workflows.argoproj.io/node-name: helloflow-8nvp9.start
    zodiac.zillowgroup.net/environment: aianalytics_stage
    zodiac.zillowgroup.net/service: aip-playground
  creationTimestamp: "2023-10-12T16:58:38Z"
  labels:
    aip.zillowgroup.net/aip-pod-default: "true"
    metaflow.org/experiment: e42
    metaflow.org/flow_name: HelloFlow
    metaflow.org/tag_e42: "true"
    metaflow.org/tag_t42: "true"
    pipelines.kubeflow.org/enable_caching: "true"
    pipelines.kubeflow.org/kfp_sdk_version: 1.0.81-1.8.10
    pipelines.kubeflow.org/pipeline-sdk-type: kfp
    sidecar.istio.io/inject: "false"
    tags.ledger.zgtools.net/ai-experiment-name: e42
    tags.ledger.zgtools.net/ai-flow-name: HelloFlow
    tags.ledger.zgtools.net/ai-step-name: start
    user: aip-playground-dev
    workflows.argoproj.io/completed: "false"
    workflows.argoproj.io/workflow: helloflow-8nvp9
    zodiac.zillowgroup.net/environment: aianalytics_stage
    zodiac.zillowgroup.net/owner: talebz
    zodiac.zillowgroup.net/product: batch
    zodiac.zillowgroup.net/service: aip-playground
    zodiac.zillowgroup.net/team: ai-platform
```

talebzeghmi committed:
AIP-7393 Add Flow labels to Workflow for ArgoUI

alexlatchford commented on metaflow/plugins/aip/aip.py:
```diff
@@ -889,7 +888,42 @@ def _set_container_labels(self, container_op: ContainerOp):
+
+    def _set_container_labels(self, container_op: ContainerOp):
+        # TODO(talebz): A Metaflow plugin framework to customize tags, labels, etc.
+        container_op.add_pod_label("aip.zillowgroup.net/aip-pod-default", "true")
```
Do we still need this label?

talebzeghmi replied:
@aaron-arellano reminded me that we replaced this with `BATCH_COMPONENT_LABEL_KEY = "workflows.argoproj.io/workflow"` https://gitlab.zgtools.net/analytics/artificial-intelligence/ai-platform/aip-infrastructure/aip-admission-webhook-service/-/blob/main/aip_admission_webhook_service/config.py#L8 

which would apply to all Argo Workflow pods!

cloudw replied:
Do we expect this line to be removed then?

talebzeghmi replied:
good point, it's no longer used!
but renaming it to `aip.zillowgroup.net/aip-wfsdk-pod` for any future use.

alexlatchford commented on metaflow/plugins/aip/aip.py:
```diff
@@ -889,7 +888,42 @@ def _set_container_labels(self, container_op: ContainerOp):
+        # Add in Zodiac service and team labels to the aip pods if the environment variable is
+        # present in the notebook (individual profile notebooks only) and set them. These labels
+        # are not being added by poddefaults as they were removed. Workflows launched in project
+        # profiles still get these labels added via poddefaults. Also adds in logging topic
```
Sorry maybe I'm not following here, are we still relying on PodDefaults or is this done via the Webhook now?

aaron-arellano replied:
this is needed to support zodiac service per notebook, that is why it is ignored in the webhook, for individual namespaces.

talebzeghmi replied:
Thanks @aaron-arellano , I'll add that to the code comments, because we forget this context!

talebzeghmi replied:
done, with Aaron's comment and ChatGPT help!

alexlatchford commented on metaflow/plugins/aip/aip.py:
```diff
@@ -859,20 +863,15 @@ def _create_volume(
-        container_op.add_pod_label("sidecar.istio.io/inject", "false")
+    def _get_flow_labels(self) -> Dict[str, str]:
+        # function return variable
+        ret: Dict[str, str] = {}
```
nit: `ret` -> `flow_labels`, maybe more descriptive?

talebzeghmi replied:
done!

alexlatchford commented on metaflow/plugins/aip/aip.py:
```diff
@@ -889,7 +888,42 @@ def _set_container_labels(self, container_op: ContainerOp):
+
+        prefix = "metaflow.org"
+        container_op.add_pod_annotation(f"{prefix}/step", container_op.name)
+        container_op.add_pod_annotation(f"{prefix}/run_id", METAFLOW_RUN_ID)
```
Classic question but seems like we have a bit of a mix between:

- Flow vs Container/Step level
- Labels vs Annotations

Is it possible to compile a quick table of what we have as an annotation versus a label and whether they are at a flow or step level? Maybe put it in the PR description so we can summarize what we have?

talebzeghmi replied:
added at the top of the PR

talebzeghmi committed:
PR feedback

cloudw commented on metaflow/plugins/aip/aip.py:
```diff
@@ -859,20 +863,15 @@ def _create_volume(
+        ret_flow_labels[f"{prefix}/flow_name"] = self.name
         if self.experiment:
-            container_op.add_pod_annotation(f"{prefix}/experiment", self.experiment)
+            ret_flow_labels[f"{prefix}/experiment"] = self.experiment
```
As we already have experiments as flow labels, do you feel we can remove experiment from flow tags, or prevent it from being duplicated in flow labels? https://github.com/zillow/metaflow/blob/tz/AIP-7393-argo-labels/metaflow/plugins/aip/aip_cli.py#L296-L298

Right now in flow label you can see experiment twice:

>   labels:
>     **metaflow.org/experiment: e42**
>     metaflow.org/flow_name: HelloFlow
>     **metaflow.org/tag_e42: "true"**
>

talebzeghmi replied:
This is because the experiment name becomes a tag, see https://github.com/zillow/metaflow/blob/9d11eded9e453ebb2f0d816d40379fd064c164ce/metaflow/plugins/aip/aip_cli.py#L298-L297

and all tags become pod labels too :)

cloudw replied:
Not going to block this PR on this minor label duplication, but experiment was a KFP concept, whereas we are not utilizing metaflow project or branches. Maybe we can find a better way to group flows while reducing the number of concepts.

talebzeghmi replied:
created https://zillowgroup.atlassian.net/browse/AIP-7769 which I'll tackle this sprint
thank you!

cloudw commented on metaflow/plugins/aip/aip.py:
```diff
@@ -901,26 +934,10 @@ def _set_container_labels(self, container_op: ContainerOp):
+            # Please see comments on how and why ZILLOW_ZODIAC_SERVICE label for more.
             container_op.add_pod_annotation(
                 "logging.zgtools.net/topic",
                 f"log.fluentd-z1.{ZILLOW_ZODIAC_SERVICE}.dev",
```
Not really part of this PR, but I'm curious why logging topic always ends with `.dev`. Is there any chance it can also be `.stage` / `.prod` etc?

talebzeghmi replied:
Nice question!:
* The intent of this code is only for Notebook launched workflows and hence dev is ok.
* However, I noticed that `logging.zgtools.net/topic` is deprecated for `logging.zgtools.net/index`,
  * see webhook code https://gitlab.zgtools.net/analytics/artificial-intelligence/ai-platform/aip-infrastructure/aip-admission-webhook-service/-/blob/main/aip_admission_webhook_service/components/pod/config_utils.py#L110
  * @aaron-arellano , do you believe this should be changed to use `logging.zgtools.net/index` ?
      * or is this a question for Max, as part of the ZAP migration? see how `logging.zgtools.net/topic` is specified by customers in `kubeflow-k8s-profile`s and not applied with the webhook: https://gitlab.zgtools.net/analytics/artificial-intelligence/ai-platform/aip-infrastructure/kubeflow-k8s-profiles/-/blob/main/profiles/prod/rental-grs-prod/rental-grs-prod.yaml#L9

freshbaby77 replied:
In ZAP, we will change these customized labels from namespace to AIPEnvironment which is ZAP specific resource in namespace level. We will also update these config_util from AIPEnvironment annotations instead of namespace. But for this edge case in rental_grs in profile about whether to use index or topic, I'll leave it to the original owner for this logic update as I'm not sure it will impact the existing rental-grs logging if we update it to` logging.zgtools.net/index`.

talebzeghmi replied:
note: updated the annotation to `"logging.zgtools.net/index",`

talebzeghmi committed:
AIP-7726 add default cards (#260)

* AIP-7726 add default cards

* param fix

* PR feedback
- comment archive none
- add card id to support current.card.append()
- add example of notebook card & multiple cards

* aip.zillowgroup.net/aip-pod-default -> aip.zillowgroup.net/aip-wfsdk-pod

talebzeghmi committed:
logging.zgtools.net/index -> logging.zgtools.net/topic

talebzeghmi committed:
fix card pathspec for nested foreach

talebzeghmi committed:
fix github test.yml branches

talebzeghmi committed:
resources_flow.py label fixes
